### PR TITLE
Finite dimensional quotients as vector spaces.

### PR DESCRIPTION
**Pull request by HechtiDerLachs**
@afkafkafk13 `vector_space` for modules over polynomial rings.

### CODE REVIEW

File: src/Modules/missing_functionality.jl [hunk 1]
```diff
@@ -77,3 +77,114 @@ end=#
...
+  R = base_ring(F)
+  # Iterate through all components until we have everything
+  result = elem_type(F)[]
+  for i in 1:ngens(F)
+    m = ideal(R, gens(R))
+    pom = ideal(R, one(R))
+    while true
+      new_mons = [x for x in gens(pom) if !(x*F[i] in lead_I)]
+      iszero(length(new_mons)) && break
+      result = vcat(result, [x*F[i] for x in new_mons])
```
Comment (fingolfin) [inline src/Modules/missing_functionality.jl:97]: This is an extremely minor thing, but I just wanted to point out that depending on the ring, `F[i]` might allocate a new object each time it is called, so perhaps pull it out before the loop? Also, `x*F[i]` is computed twice for all remaining in `new_mons`. 
```suggestion
    Fi = F[i]
    m = ideal(R, gens(R))
    pom = ideal(R, one(R))
    while true
      new_mons = [x*Fi for x in gens(pom)]
      filter!(x -> !(x in lead_I), new_mons)
      ismpty(new_mons) && break
      result = vcat(result, new_mons)
```

File: src/Modules/missing_functionality.jl [hunk 2]
```diff
@@ -77,3 +77,114 @@ end=#
     
 subquo_type(::Type{RingType}) where {RingType<:Ring} = SubquoModule{elem_type(RingType)}
 subquo_type(R::RingType) where {RingType<:Ring} = subquo_type(typeof(R))
+
+function _kbase(I::SubModuleOfFreeModule{<:MPolyRingElem{<:FieldElem}};
+    ordering::ModuleOrdering=default_ordering(ambient_free_module(I))
+  )
+  lead_I = leading_module(I, ordering)
+  @show ordering
+  @show lead_I
```
Comment (fingolfin) [inline src/Modules/missing_functionality.jl:86]: I guess this means this code is not yet ready? Then perhaps mark it as draft?

File: src/Modules/missing_functionality.jl (hunk 1 above)
Comment (fingolfin) [inline src/Modules/missing_functionality.jl:97]: This is an extremely minor thing, but I just wanted to point out that depending on the ring, `F[i]` might allocate a new object each time it is called, so perhaps pull it out before the loop? Also, `x*F[i]` is computed twice for all remaining in `new_mons`. 
```suggestion
    Fi = F[i]
    m = ideal(R, gens(R))
    pom = ideal(R, one(R))
    while true
      new_mons = [x*Fi for x in gens(pom)]
      filter!(x -> !(x in lead_I), new_mons)
      ismpty(new_mons) && break
      result = vcat(result, new_mons)
```

**Comment by fingolfin** [comment]
Does the geometry team have a plan who will review this PR?

**Comment by afkafkafk13** [comment]
> Does the geometry team have a plan who will review this PR?

I am planning to do this and I was already in contact with @HechtiDerLachs about certain details.

**Comment by fingolfin** [comment]
Excellent. I've assigned the PR to you then to reflect this in the webinterface :-)

**Comment by afkafkafk13** [comment]
@HechtiDerLachs :  Could you check whether some of the problems you needed to avoid are fixed by now? There is a vector_space_basis for quo's by now in the ideal and module setting.

File: test/Modules/module-localizations.jl [hunk 3]
```diff
@@ -131,3 +131,20 @@ end
...
 end
+
+@testset "as vector spaces" begin
+
+R, (x, y) = QQ["x", "y"]
+R, _ = localization(R, complement_of_point_ideal(R, [0, 1]))
+F2 = free_module(R, 2)
+I, inc_I = sub(F2, [x^3*F2[1], (y-1)*y*F2[1], x^2*F2[1]-(y-1)*F2[2], x*F2[2], (y-1)^2*F2[2]])
+M, pr_M = quo(F2, I)
+# tests below should run one day.
```
Comment (afkafkafk13) [inline test/Modules/module-localizations.jl:142]: can you outline, where the problems are?

Comment (HechtiDerLachs) [reply test/Modules/module-localizations.jl:142]: I don't remember in full detail. I think the map `V -> M` works. But `M -> V` needs to bring elements in `M` to normal form and then collect coefficients. Even in the case of trivial denominators, there was something missing from the module side (was it messing with the local orderings...?). Another step would then be to also allow non-trivial denominators.

Comment (afkafkafk13) [reply test/Modules/module-localizations.jl:142]: ah, yes.
Reduced normal forms for localized rings do not make sense in the general setting (will not terminate), but in the present situation (with a weighted local ordering and in the presence of a highest corner) they exist and can be computed. This is, what you will need here.
But given the fact, that even for the reduced normal form of an element (not ideal!) in the global case a hack was necessary, I see things that are more urgent. 
(The denominators are then nothing to worry about, as soon as the rest works)

**Comment by simonbrandhorst** [comment]
What is the status of this PR?

**Comment by fingolfin** [comment]
Note that we have a function `monomial_basis`, and also a function `vector_space_basis`. And now this here. It would be good to make sure we are not reinventing the wheel everywhere, and that these things fit together...

**Comment by wdecker** [comment]
@fingolfin As I can see from the docstrings in UngradedModules.jl, the `vector_space_basis` function also
returns a monomial basis in each case so that this function could be renamed to `monomial_basis`.
Similarly for `_kbase`. We should discuss, however, whether we should add an additional argument for the monomial ordering to be used whenever this is appropriate. @jankoboehm @HechtiDerLachs @afkafkafk13 ???

File: src/Modules/missing_functionality.jl [hunk 4]
```diff
@@ -77,3 +77,114 @@ end=#
...
+      new_mons = [x for x in gens(pom) if !(x*F[i] in lead_I)]
+      iszero(length(new_mons)) && break
+      result = vcat(result, [x*F[i] for x in new_mons])
+      pom = pom * m
+    end
+  end
+  return result
+end
+
+function vector_space(
```
Comment (fingolfin) [inline src/Modules/missing_functionality.jl:104]: A comment (or even better, docstring) that explains what this function does (what does it take, what does it return) would be helpful

**Comment by fingolfin** [comment]
So what exactly needs to be done here? Maybe next week some people can discuss this in person in Saarbrücken?

**Comment by fingolfin** [comment]
@afkafkafk13 mentions that one blocked is the lack of "strong normal forms in the local case" which apparently are needed to write down the isomorphism. But apparently it is not that hard to implement, so "someone" just has to do it... But again, maybe it can be discussed next week (and details added here)

**Comment by simonbrandhorst** [comment]
@HechtiDerLachs is this still relevant? Can this be closed?

**Comment by HechtiDerLachs** [comment]
I forgot what all this was really about and why it didn't proceed. But somewhere local orderings and modules pop up and I don't have the appetite to get into any of these things right now. The name of the PR suggests that we wanted this functionality for our School in Lille last year. Since it has now already taken place and we seem to have managed without this stuff, I would be OK with throwing it away. It can still be digged out if anyone ever wanted to come back to this.

**Pull request closed without merging**